design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,260 @@
+can be broken between major releases. It is worth noting that when a
+new major release is introduced, the previous one is still maintained
+and supported, something that does not exist with minor releases.
+
+#### The Deprecation Flow (for Minor releases)
+As stated before, only features that have not yet been graduated
+can be discontinued between minor releases.
+Therefore, a feature can be deprecated in Alpha and Beta stages only.
+
+These are the steps needed to deprecate a feature:
+- Proposal: Prepare a proposal to discontinue a feature with proper
+  reasoning, phases, exact timelines and functional alternatives (if any).
+  The proposal should be reviewed and approved.
+- Notification: Notify the project community about the feature
+  discontinuation based on the approved proposal.
+  All details of the plan should be provided to allow users and possibly
+  d/s projects to adjust.
+  Use all community media options to distribute this information
+  (e.g. mailing list, slack channel, community meetings).
+- Deprecation warnings: Add deprecation warnings at runtime to warn users
+  that the feature is planned to be discontinued.
+  Warnings should be raised when:
+  - Feature API fields are accessed.
+  - Feature FG is activated.
+  - Behavior related to the feature is detected (optional).
+- Removal: Feature removal involves dropping the core functionality
+  of a feature and its exposed API.
+  - The core implementation can be dropped in two steps:
+    - The FG is dropped by assuring it is never reported as set
+      (i.e. even if it is left by the operator configured, internally
+       it is ignored).
+      At this stage, the core implementation will follow the FG conditions
+      and therefore from the outside the feature is inactive.
+    - In case there are no side effects, the core implementation code can
+      be removed.
+  - The API types are not to be removed, as it may have implications
+    with the underlying storage which has already persisted them.
+    Kubernetes has not removed fields, it just kept them around with the
+    warning that they have been deprecated and no longer available.
+
+    While keeping fields around for a period of a release or two makes
+    sense, beyond a limited period it adds a burden on dragging leftover
+    fields around to eternity.
+
+> **Note**: The only reference seen on why fields should not be removed
+> was mentioned [here](https://github.com/kubernetes/kubernetes/issues/52185).
+> But it is unclear if this is relevant for Alpha stage features.
+
+### Exceptions
+While the project strives to maintain a stable contract with its users,
+there may be scenarios where the policy describe here will not be a fit.
+
+Therefore, it should be acceptable to have exceptions from time to time
+given a very good reasoning and an agreement from 2/3 of the project
+maintainers.
+
+## Implementation Phases
+N/A
+
+## Open Issues
```

Comment:
Additional possible open issues for discussion:
- Can Alpha features get introduced in z versions?
- CI & e2e restrictions:
   - Run e2e tests of Alpha features in their own separate job on the latest k8s version, non gating.
   - Monitor Alpha features e2e execution results and use the data as a gating (e.g. 95% success rate) to pass it to Beta stage.

Reply:
Alpha in zstreams: No. zstreams are just for stability. Exceptions confirm the norm.

Blocking CI for Alpha: No. Alpha fetaures should not gate main. But it sounds like Beta features should.

Reply:
Done